obspy/io/gcf/libgcf.py

```diff
@@ -0,0 +1,159 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+from future.builtins import *  # NOQA
+
+import numpy as np
+
+from obspy import UTCDateTime
+
+sps_d = {  # Table 3.1: special sample rates
+    157: 0.1,
+    161: 0.125,
+    162: 0.2,
+    164: 0.25,
+    167: 0.5,
+    171: 400,
+    174: 500,
+    176: 1000,
+    179: 2000,
+    181: 4000}
+time_offsets_d = {  # Table 3.1: Time fractional offset denominator
+    171: 8.,
+    174: 2.,
+    176: 4.,
+    179: 8.,
+    181: 16.}
+compression_d = {  # Table 3.2: format field to data type
+    1: '>i4',
+    2: '>i2',
+    4: '>i1'}
+bits_d = {  # Table 3.2: format field to data bytes
+    1: 4,  # 32bits
+    2: 2,  # 16bits
+    4: 1}   # 8bits
+
+
+def is_gcf(f):
+    """
+    Test if file is GCF by reading at least 1 data block
+    """
+    header, data = read_data_block(f)
+
+
+def decode36(data):
+    """
+    Decode base 36 data
+    """
+    # http://geophysics.eas.gatech.edu/GTEQ/Scream4.4/Decoding_Base_36_numbers_C.htm
+    s = ''
+    while data:
+        imed = data % 36
+        if imed > 9:
+            imed += 7
+        s = chr(imed + 48) + s
+        data = np.uint32(data / 36)
+    return s
+
+
+def decode_date_time(data):
+    """
+    Decode date and time field.
+
+    The date code is a 32 bit value specifying the start time of the block.
+    Bits 0-16 contain the number of seconds since midnight,
+    and bits 17-31 the number of days since 17th November 1989.
+    """
+    days = data >> 17
+    secs = data & 0x1FFFF
+    starttime = UTCDateTime('1989-11-17') + days * 86400 + secs
+    return starttime
+
+
+def read_data_block(f, headonly=False, channel_prefix="HH", **kwargs):
+    """Read one data block from GCF file.
+
+    more details can be found here:
+    http://geophysics.eas.gatech.edu/GTEQ/Scream4.4/GCF_Specification.htm
+    f - file object to read from
+    if skipData is True, Only header is returned.
+    if not a data block (SPS=0) - returns None.
+    """
+    # get ID
+    sysid = f.read(4)
+    if not sysid:
+        raise EOFError  # got to EOF
+    sysid = np.frombuffer(sysid, count=1, dtype='>u4')
+    if sysid >> 31 & 0b1 > 0:
+        sysid = (sysid << 6) >> 6
+    sysid = decode36(sysid)
+    # get Stream ID
+    stid = np.frombuffer(f.read(4), count=1, dtype='>u4')
+    stid = decode36(stid)
+    # get Date & Time
+    data = np.frombuffer(f.read(4), count=1, dtype='>u4')
+    starttime = decode_date_time(data)
+    # get data format
+    # get reserved, SPS, data type compression, number of 32bit records (N)
+    reserved, sps, compress, N = np.frombuffer(f.read(4), count=4, dtype='>u1')
+    compression = compress & 0b00001111  # get compression code
+    t_offset = compress >> 4  # get time offset
+    if t_offset > 0:
+        starttime = starttime + t_offset / time_offsets_d[sps]
+    if sps in sps_d:
+        sps = sps_d[sps]  # get special SPS value if needed
```

Comment:
I don't think is't possible if file is indeed an original GCF format:
The value of the sample rate field (byte 13) generally indicates the
number of samples per second. However, as it is an unsigned 8-bit
integer, it is not possible to directly represent sample rates higher than
255sps or lower than 1sps. Therefore, some unlikely values have been
used to represent such sample rates. These are shown in sps_d.

Source: [http://www.guralp.com/apps/ok?doc=GCF_format](http://www.guralp.com/apps/ok?doc=GCF_format) (page 8)